purchase_request/models/purchase_request.py

```diff
@@ -0,0 +1,251 @@
+import openerp.addons.decimal_precision as dp
+
+_STATES = [
+    ('draft', 'Draft'),
+    ('to_approve', 'To be approved'),
+    ('approved', 'Approved'),
+    ('rejected', 'Rejected')
+]
+
+
+class PurchaseRequest(models.Model):
+
+    _name = 'purchase.request'
+    _description = 'Purchase Request'
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+
+    @api.model
+    def _company_get(self):
+        company_id = self.env['res.company']._company_default_get(self._name)
+        return self.env['res.company'].browse(company_id.id)
+
+    @api.model
+    def _get_default_requested_by(self):
+        return self.env['res.users'].browse(self.env.uid)
+
+    @api.model
+    def _get_default_name(self):
+        return self.env['ir.sequence'].get('purchase.request')
+
+    @api.model
+    def _default_picking_type(self):
+        type_obj = self.env['stock.picking.type']
+        company_id = self.env.context.get('company_id') or \
+            self.env.user.company_id.id
+        types = type_obj.search([('code', '=', 'incoming'),
+                                 ('warehouse_id.company_id', '=', company_id)])
+        if not types:
+            types = type_obj.search([('code', '=', 'incoming'),
+                                     ('warehouse_id', '=', False)])
+        return types[:1]
+
+    @api.multi
+    @api.depends('name', 'origin', 'date_start',
+                 'requested_by', 'assigned_to', 'description', 'company_id',
+                 'line_ids', 'picking_type_id')
+    def _compute_is_editable(self):
+        for rec in self:
+            if rec.state in ('to_approve', 'approved', 'rejected'):
+                rec.is_editable = False
+            else:
+                rec.is_editable = True
+
+    @api.multi
+    def _track_subtype(self, init_values):
+        for rec in self:
+            if 'state' in init_values and rec.state == 'to_approve':
+                return 'purchase_request.mt_request_to_approve'
+            elif 'state' in init_values and rec.state == 'approved':
+                return 'purchase_request.mt_request_approved'
+            elif 'state' in init_values and rec.state == 'rejected':
+                return 'purchase_request.mt_request_rejected'
+        return super(PurchaseRequest, self)._track_subtype(init_values)
+
+    name = fields.Char('Request Reference', size=32, required=True,
+                       default=_get_default_name,
+                       track_visibility='onchange')
+    origin = fields.Char('Source Document', size=32)
+    date_start = fields.Date('Creation date',
+                             help="Date when the user initiated the "
+                                  "request.",
+                             default=fields.Date.context_today,
+                             track_visibility='onchange')
+    requested_by = fields.Many2one('res.users',
+                                   'Requested by',
+                                   required=True,
```

Comment:
@JonathanNEMRY :-1: deleting a user must raise an error, not delete everything the user has done.